src/opencl/mod.rs

```diff
@@ -130,28 +135,38 @@ impl Device {
     pub fn by_bus_id(bus_id: BusId) -> GPUResult<Device> {
         Device::all().and_then(|devs| {
             devs.into_iter()
-                .find(|d| d.bus_id == bus_id)
+                .find(|d| match d.bus_id {
+                    Some(id) => bus_id == id,
+                    None => false,
+                })
                 .ok_or(GPUError::DeviceNotFound)
         })
     }
 
     pub fn by_brand(brand: Brand) -> GPUResult<Vec<Device>> {
         match utils::find_platform(brand.platform_name())? {
-            Some(plat) => ocl::Device::list_all(plat)?
-                .into_iter()
-                .map(|d| {
-                    (|| -> GPUResult<Device> {
-                        Ok(Device {
-                            brand,
-                            name: d.name()?,
-                            memory: get_memory(d)?,
-                            bus_id: brand.extract_bus_id(d)?,
-                            platform: plat,
-                            device: d,
-                        })
-                    })()
-                })
-                .collect(),
+            Some(plat) => {
+                // List all GPU devices on `plat`.
+                ocl::Device::list(plat, Some(ocl::core::DeviceType::GPU))?
+                    .into_iter()
+                    .filter(|d| {
+                        // Only return available devices.
+                        d.is_available().unwrap_or(false)
+                    })
+                    .map(|d| {
+                        (|| -> GPUResult<Device> {
```

Comment:
What akward construction is this inner lambda? This really should be removed

Reply:
Looks like a vestigial type hack. Removed.